android/examples/todo-sync/src/main/java/com/mongodb/stitch/android/examples/todosync/TodoAdapter.java

```diff
@@ -69,19 +79,68 @@ public int getItemCount() {
     return todoItems.size();
   }
 
-  public void updateItems(final Task<List<TodoItem>> todoItems) {
-    todoItems.addOnCompleteListener(task -> {
+  public synchronized void updateOrAddItem(final TodoItem todoItem) {
+    if (todoItemIdsToIdx.containsKey(todoItem.getId())) {
+      todoItems.set(todoItemIdsToIdx.get(todoItem.getId()), todoItem);
+    } else {
+      todoItems.add(todoItem);
+      todoItemIdsToIdx.put(todoItem.getId(), todoItems.size() - 1);
+    }
+    TodoAdapter.this.notifyDataSetChanged();
+  }
+
+  public synchronized void removeItemById(final ObjectId id) {
+    if (!todoItemIdsToIdx.containsKey(id)) {
+      return;
+    }
+    todoItems.remove((int) todoItemIdsToIdx.remove(id));
+    // update indexes
+    for (int i = 0; i < todoItems.size(); i++) {
+      final TodoItem item = todoItems.get(i);
+      todoItemIdsToIdx.put(item.getId(), i);
+    }
+    TodoAdapter.this.notifyDataSetChanged();
+  }
+
+  public synchronized void replaceItems(
+      final Task<List<TodoItem>> newTodoItems, final boolean removeUnseen) {
+    newTodoItems.addOnCompleteListener(task -> {
       if (!task.isSuccessful()) {
         Log.e(TAG, "failed to get items", task.getException());
+        return;
+      }
+      final Set<Integer> unseenIdxs = new HashSet<>(todoItemIdsToIdx.values());
+      final List<TodoItem> newItems = task.getResult();
+
+      // replace existing items and add new ones
+      for (final TodoItem newItem : newItems) {
```

Comment:
Interesting that this is a road a user might have to go down. We may want to consider (in the far future) simplifying processes like this for users.

Reply:
I think this is just crappy application design on my part and not using android adapters well. I think Room improves this.